### PR TITLE
postgres server backup-support

### PR DESCRIPTION
**Pull request by karisN**
add support for backup redundancy options

##### SUMMARY
Postgres server has backup redundancy options, which can only be set during the creation. These parameters are missing from the plugin today.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
azure_rm_postgresqlserver

##### ADDITIONAL INFORMATION
Add: 
geo redundant backup flag
backup retention days
creation modes (default, point in time restore, geo redundant restore)

### CODE REVIEW

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 1]
```diff
@@ -239,20 +272,39 @@ def exec_module(self, **kwargs):
                 elif key == "location":
                     self.parameters["location"] = kwargs[key]
                 elif key == "storage_mb":
-                    self.parameters.setdefault("properties", {}).setdefault("storage_profile", {})["storage_mb"] = kwargs[key]
+                    self.parameters.setdefault("properties", {}).setdefault("storage_profile", {})["storage_mb"] = \
+                    kwargs[key]
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py]: continuation line missing indentation or outdented

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 2]
```diff
@@ -51,6 +51,14 @@
         description:
             - The maximum storage allowed for a server.
         type: int
+    geo_redundant_backup:
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py:58]: Missing default value in document!

File: plugins/modules/azure_rm_postgresqlserver_info.py [hunk 3]
```diff
@@ -245,6 +245,8 @@ def format_item(self, item):
             'admin_username': d['administrator_login'],
             'user_visible_state': d['user_visible_state'],
             'fully_qualified_domain_name': d['fully_qualified_domain_name'],
+            'geo_redundant_backup': d['storage_profile']['geo_redundant_backup'],
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver_info.py:257]: @karisN  Please add follow content to return value!
```e.g.
        backup_retention_days:
            description:
                - Backup retention period between 7 and 35 days. 7 days by default if not set.
            returned: always
            type: int
            sample: 7
        geo_redundant_backup:
            description:
                - Choose between locally redundant(default) or geo-redundant backup. This cannot be updated after first deployment.
            returned: always
            type: str
            sample: Disabled

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 4]
```diff
@@ -192,7 +218,14 @@ def __init__(self):
             ),
             create_mode=dict(
                 type='str',
-                default='Default'
+                default='default',
+                choices=['default', 'geo_restore', 'point_in_time_restore']
+            ),
+            source_server_id=dict(
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py:236]: @karisN What do these two parameters do “ source_server_id" and "restore_point_in_time"? Can you help to add a use case for these two parameters? Thank you very much!

Comment (karisN) [reply plugins/modules/azure_rm_postgresqlserver.py:236]: I updated the docs a bit more, hopefully that makes sense now.

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 5]
```diff
@@ -51,6 +51,15 @@
         description:
             - The maximum storage allowed for a server.
         type: int
+    geo_redundant_backup:
+        description:
+            - Choose between locally redundant(default) or geo-redundant backup. This cannot be updated after first deployment
+        type: bool
+        default: default
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py]: Argument 'geo_redundant_backup' in documentation defines default as C(Disabled) .
```suggestion
        type: str
        default: Disabled
        choices:
            - Disabled
            - Enabled
```

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 6]
```diff
@@ -77,8 +86,19 @@
...
-        default: Default
+            - Create mode of SQL Server. Blank (default), restore from geo redundant (geo_restore), or restore from point in time (point_in_time_restore).
+        default: default
+        choices:
+            - default
+            - geo_restore
+            - point_in_time_restore
+    source_server_id:
+        description:
+            - Id if the source server if C(create_mode) is not I(default).
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py]:
```suggestion
            - Id if the source server if I(create_mode=default).
```

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 7]
```diff
@@ -77,8 +86,19 @@
...
+    restore_point_in_time:
+        description:
+            - Restore point creation time (ISO8601 format), specifying the time to restore from.
+            - Required if C(create_mode) is set to I(point_in_time_restore).
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py]:
```suggestion
            - Required if I(create_mode=point_in_time_restore).
```

**Comment by Fred-sun** [comment]
@karisN Would you help to add test case for these paramter (source_server_id and restore_point_in_time) . Thank you very much!

**Comment by karisN** [comment]
> @karisN Would you help to add test case for these paramter (source_server_id and restore_point_in_time) . Thank you very much!

I'm not actually familiar with how the tests are setup in this project, but AFAIK geo backup occurs only twice a day. It's not likely a backup will become available for restore immediately in a test case, additionally I don't have info on when point_in_time restore is available after creation either. What are my options for restore test cases?

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 8]
```diff
@@ -178,6 +197,13 @@ def __init__(self):
             storage_mb=dict(
                 type='int'
             ),
+            geo_redundant_backup=dict(
+                type='bool',
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py:213]:
```suggestion
                type='str',
```

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 9]
```diff
@@ -178,6 +197,13 @@ def __init__(self):
...
+                default=False
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py:214]:
```suggestion
                default=Disabled,
                choices=['Disabled', 'Enabled']
```

File: plugins/modules/azure_rm_postgresqlserver.py (hunk 5 above)
Comment (Fred-sun) [reply plugins/modules/azure_rm_postgresqlserver.py]: @karisN Please keep consistent with the argument spec defined. Thank you very much!

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 10]
```diff
@@ -178,6 +197,14 @@ def __init__(self):
             storage_mb=dict(
                 type='int'
             ),
+            geo_redundant_backup=dict(
+                type='str',
+                default=Disabled,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py]:
```suggestion
                default='Disabled',
```

**Review by Fred-sun — COMMENTED** [review]
@karisN Please delete 252 line. the file " tests/sanity/ignore-2.10.txt". You have resolved.
plugins/modules/azure_rm_postgresqlserver.py validate-modules:parameter-type-not-in-doc

**Comment by Fred-sun** [comment]
@karisN Could you please help to authorize me to update this PR? Thank you very much!

**Comment by dpodder** [comment]
Hi @Fred-sun and team, is there anything holding up this PR? We were looking for similar requirements in another project and this would fill the exact missing functionality we'd need.

**Comment by Fred-sun** [comment]
@dpodder We plan to merge this PR next week, and I will push forward the solution. Thank you very much!

File: plugins/modules/azure_rm_postgresqlserver.py [hunk 11]
```diff
@@ -72,17 +86,34 @@
...
+            - Create mode of SQL Server. Blank (default), restore from geo redundant (geo_restore), or restore from point in time (point_in_time_restore).
+        type: str
+        default: default
+        choices:
+            - default
+            - geo_restore
+            - point_in_time_restore
+    source_server_id:
+        description:
+            - Id if the source server if I(create_mode=default).
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_postgresqlserver.py:105]: @karisN There is something wrong with this description statement. Please refer to the relevant documents of this module and describe it again. Thank you very much!

**Pull request merged**